mixnet/node.py

```diff
@@ -1,107 +1,159 @@
-    X25519PrivateKey,
-)
 from pysphinx.sphinx import (
     Payload,
     ProcessedFinalHopPacket,
     ProcessedForwardHopPacket,
     SphinxPacket,
-    UnknownHeaderTypeError,
 )
 
-from mixnet.config import MixNodeConfig, NodeAddress
-from mixnet.poisson import poisson_interval_sec
+from mixnet.config import GlobalConfig, NodeConfig
+from mixnet.connection import DuplexConnection, MixSimplexConnection
+from mixnet.gossip import GossipChannel
+from mixnet.packet import Fragment, MessageFlag, MessageReconstructor, PacketBuilder
 
-PacketQueue: TypeAlias = "asyncio.Queue[Tuple[NodeAddress, SphinxPacket]]"
-PacketPayloadQueue: TypeAlias = (
-    "asyncio.Queue[Tuple[NodeAddress, SphinxPacket | Payload]]"
-)
+BroadcastChannel: TypeAlias = asyncio.Queue[bytes]
 
 
-class MixNode:
+class Node:
     """
-    A class handling incoming packets with delays
-
-    This class is defined separated with the MixNode class,
-    in order to define the MixNode as a simple dataclass for clarity.
+    This represents any node in the network, which:
+    - generates/gossips mix messages (Sphinx packets)
+    - performs cryptographic mix (unwrapping Sphinx packets)
+    - generates noise
     """
 
-    config: MixNodeConfig
-    inbound_socket: PacketQueue
-    outbound_socket: PacketPayloadQueue
-    task: asyncio.Task  # A reference just to prevent task from being garbage collected
-
-    @classmethod
-    async def new(
-        cls,
-        config: MixNodeConfig,
-    ) -> Self:
-        self = cls()
+    config: NodeConfig
+    global_config: GlobalConfig
+    mixgossip_channel: GossipChannel
+    reconstructor: MessageReconstructor
+    broadcast_channel: BroadcastChannel
+    packet_size: int
+
+    def __init__(self, config: NodeConfig, global_config: GlobalConfig):
         self.config = config
-        self.inbound_socket = asyncio.Queue()
-        self.outbound_socket = asyncio.Queue()
-        self.task = asyncio.create_task(self.__run())
-        return self
+        self.global_config = global_config
+        self.mixgossip_channel = GossipChannel(config.gossip, self.__process_msg)
+        self.reconstructor = MessageReconstructor()
+        self.broadcast_channel = asyncio.Queue()
 
-    async def __run(self):
-        """
-        Read SphinxPackets from inbound socket and spawn a thread for each packet to process it.
+        sample_packet, _ = PacketBuilder.build_real_packets(
+            bytes(1), global_config.membership, self.global_config.max_mix_path_length
+        )[0]
+        self.packet_size = len(sample_packet.bytes())
 
-        This thread approximates a M/M/inf queue.
+    async def __process_msg(self, msg: bytes) -> bytes | None:
+        """
+        A handler to process messages received via gossip channel
+        """
+        flag, msg = Node.__parse_msg(msg)
+        match flag:
+            case MsgType.NOISE:
+                # Drop noise packet
+                return None
+            case MsgType.REAL:
```

Comment:
Hmm, if this is the node logic for processing messages received via gossip channel then the REAL message handler should do 2 things.
1) The handler should immediately pass the real message to the gossip routine, where it should be added to connection queues that have not seen the message.
2) Immediately after the above (or even concurrently) the handler should pass the real message to the sphinx processing routine, which after processing will add the output message to the queue or broadcast it.

In here I see that we are only processing the message in the sphinx routine but I don't see the gossip part. Can you point me to the place that we are doing the 1) part?

Reply:
Good point. The part 1) can be found [here](https://github.com/logos-co/nomos-specs/blob/871215742984bface1878b943c9d8f7506a10351/mixnet/node.py#L80-L80). The `__process_sphinx_packet` returns the original packet as it is, so that it can be gossiped through. But, I agree that it's not readable. Also, as you mentioned, the part 1) must be done by the gossip channel level, not in the node logic level. When imagining the libp2p gossipsub, the libp2p will anyway gossip all incoming messages (unless they're duplicate) regardless of the node internal logic. Also, that's why the noise packet needs to be dropped in the gossip channel level (, not the node logic level), to ensure that any noise isn't gossiped further. I accidentally moved this noise handling into the node logic when I refactored the code. I'll fix this right now and ping you again.

Reply:
010733c I refactored the Nomssip v.s. Node encapsulation to clarify the part 1) and 2). Now, the logic flow in the source code is very clear.
The part 1) is done in the nomssip level. The part 2) is done in the node logic level.

Reply:
Looks good. One thing to consider, have noticed that we are using `gossip` and `__gossip`, wouldn't it be more readable to use something more descriptive? Like `gossip_new` for gossiping new messages or `gossip_relay` for gossiping received messages? It's totally a nit, and I'm fine with just gossip as we have comments :)

Reply:
I agree. I just made it more readable by making `__gossip` to accept only `FlaggedPacket` object that I just defined. (also, renamed the function)